recommender/engine.go

```diff
@@ -36,28 +59,28 @@ type ClusterRecommendationReq struct {
 }
 
 type ClusterRecommendationResp struct {
-	Provider  string     `json:provider`
+	Provider  string     `json:"provider"`
 	Zones     []string   `json:"zones,omitempty"`
-	NodePools []NodePool `json:nodePools`
+	NodePools []NodePool `json:"nodePools"`
 }
 
 type NodePool struct {
-	VmType   VirtualMachine `json:vm`
-	SumNodes int            `json:sumNodes`
-	VmClass  string         `json:vmClass`
+	VmType   VirtualMachine `json:"vm"`
+	SumNodes int            `json:"sumNodes"`
+	VmClass  string         `json:"vmClass"`
 }
 
 type VirtualMachine struct {
-	Type          string  `json:type`
-	AvgPrice      float64 `json:avgPrice`
-	OnDemandPrice float64 `json:onDemandPrice`
-	Cpus          float64 `json:cpusPerVm`
-	Mem           float64 `json:memPerVm`
-	Gpus          float64 `json:gpusPerVm`
+	Type          string  `json:"type"`
+	AvgPrice      float64 `json:"avgPrice"`
+	OnDemandPrice float64 `json:"onDemandPrice"`
+	Cpus          float64 `json:"cpusPerVm"`
+	Mem           float64 `json:"memPerVm"`
+	Gpus          float64 `json:"gpusPerVm"`
 	// i/o, network
 }
 
-func (v VirtualMachine) getAttrValue(attr string) float64 {
+func (v *VirtualMachine) getAttrValue(attr string) float64 {
```

Comment:
If `attr` can pick up only a well defined set of values (Cpu, Memory) than probable we need a validation function and return with error if unsupported attribute is passed

Reply:
+1

Reply:
attribute-names are passed through a succession of methods, many of them fail incase the attribute is not supported. I agree that once the api gets cleaner we can add some generic (semantic) validation - the easiest would be to have properly annotated "domain" structs so that validation could be done out of the box with the `validator` library